src/v/cluster/rm_group_frontend.cc

```diff
@@ -45,4 +45,401 @@ rm_group_frontend::rm_group_frontend(
+            return cp.commit_group_tx(
+              commit_group_tx_request{
+                .pid = pid,
+                .tx_seq = tx_seq,
+                .group_id = group_id,
+                .timeout = timeout},
+              rpc::client_opts(model::timeout_clock::now() + timeout));
+        })
+      .then(&rpc::get_ctx_data<commit_group_tx_reply>)
+      .then([](result<commit_group_tx_reply> r) {
+          if (r.has_error()) {
+              vlog(
+                clusterlog.warn, "got error {} on remote commit tx", r.error());
+              return commit_group_tx_reply{.ec = tx_errc::timeout};
+          }
+
+          return r.value();
+      });
+}
+
+ss::future<commit_group_tx_reply> rm_group_frontend::do_commit_group_tx(
+  kafka::group_id group_id,
+  model::producer_identity pid,
+  model::tx_seq tx_seq,
+  model::timeout_clock::duration timeout) {
+    commit_group_tx_request req;
+    req.group_id = group_id;
+    req.pid = pid;
+    req.tx_seq = tx_seq;
+    req.timeout = timeout;
+
+    co_return co_await _group_router.local().commit_tx(std::move(req));
+}
+
+ss::future<abort_group_tx_reply> rm_group_frontend::abort_group_tx(
+  kafka::group_id group_id,
+  model::producer_identity pid,
+  model::timeout_clock::duration timeout) {
+    auto ntp_opt = _coordinator_mapper.local().ntp_for(group_id);
+    if (!ntp_opt) {
+        vlog(clusterlog.warn, "can't find ntp for {} ", group_id);
+        co_return abort_group_tx_reply{.ec = tx_errc::partition_not_exists};
+    }
+    auto ntp = ntp_opt.value();
+
+    auto nt = model::topic_namespace(ntp.ns, ntp.tp.topic);
+    if (!_metadata_cache.local().contains(nt, ntp.tp.partition)) {
+        vlog(clusterlog.warn, "can't find meta info for {}", ntp);
+        co_return abort_group_tx_reply{.ec = tx_errc::partition_not_exists};
+    }
+
+    auto leader_opt = _leaders.local().get_leader(ntp);
+    if (!leader_opt) {
+        vlog(clusterlog.warn, "can't find a leader for {}", ntp);
+        co_return abort_group_tx_reply{.ec = tx_errc::leader_not_found};
+    }
+    auto leader = leader_opt.value();
+    auto _self = _controller->self();
+
+    if (leader == _self) {
+        co_return co_await do_abort_group_tx(group_id, pid, timeout);
+    }
+
+    vlog(
+      clusterlog.trace,
+      "dispatching begin group tx to {} from {}",
+      leader,
+      _self);
+    co_return co_await dispatch_abort_group_tx(leader, group_id, pid, timeout);
+}
+
+ss::future<abort_group_tx_reply> rm_group_frontend::dispatch_abort_group_tx(
+  model::node_id leader,
+  kafka::group_id group_id,
+  model::producer_identity pid,
+  model::timeout_clock::duration timeout) {
+    return _connection_cache.local()
+      .with_node_client<cluster::tx_gateway_client_protocol>(
+        _controller->self(),
+        ss::this_shard_id(),
+        leader,
+        timeout,
+        [group_id, pid, timeout](tx_gateway_client_protocol cp) {
```

Comment:
keep in mind opportunities for moving. here group_id is copied into the capture, and then copied again below into the request. in both cases it looks like it could be moved. some of the group-ids I've seen from clients have been pretty big strings some prefix and uuids etc... I think this applies to the other dispatch methods too.